control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -934,40 +933,50 @@ func getHealthCheckStatusReason(healthCheckStatus, podName, podNamespace string)
 // has addresses, it will only deregister instances not in the map.
-func (r *Controller) deregisterService(apiClient *api.Client, k8sSvcName, k8sSvcNamespace string, endpointsAddressesMap map[string]bool) error {
+func (r *Controller) deregisterService(
+	ctx context.Context,
+	apiClient *api.Client,
+	k8sSvcName string,
+	k8sSvcNamespace string,
+	endpointsAddressesMap map[string]bool) (time.Duration, error) {
+
 	// Get services matching metadata from Consul
 	serviceInstances, err := r.serviceInstances(apiClient, k8sSvcName, k8sSvcNamespace)
 	if err != nil {
 		r.Log.Error(err, "failed to get service instances", "name", k8sSvcName)
-		return err
+		return 0, err
 	}
 
 	var errs error
+	var requeueAfter time.Duration
 	for _, svc := range serviceInstances {
 		// We need to get services matching "k8s-service-name" and "k8s-namespace" metadata.
 		// If we selectively deregister, only deregister if the address is not in the map. Otherwise, deregister
 		// every service instance.
 		var serviceDeregistered bool
-		if endpointsAddressesMap != nil {
-			if _, ok := endpointsAddressesMap[svc.ServiceAddress]; !ok {
-				// If the service address is not in the Endpoints addresses, deregister it.
-				r.Log.Info("deregistering service from consul", "svc", svc.ServiceID)
-				_, err := apiClient.Catalog().Deregister(&api.CatalogDeregistration{
-					Node:      svc.Node,
-					ServiceID: svc.ServiceID,
-					Namespace: svc.Namespace,
-				}, nil)
-				if err != nil {
-					// Do not exit right away as there might be other services that need to be deregistered.
-					r.Log.Error(err, "failed to deregister service instance", "id", svc.ServiceID)
-					errs = multierror.Append(errs, err)
-				} else {
-					serviceDeregistered = true
-				}
+
+		if addressIsMissingFromEndpointsMap(svc.ServiceAddress, endpointsAddressesMap) {
+			// If graceful shutdown is enabled, continue to the next service instance and
+			// mark that an event requeue is needed. We should requeue at the longest time interval
+			// to prevent excessive re-queues. Also, updating the health status in Consul to Critical
+			// should prevent routing during gracefulShutdown.
+			podShutdownDuration, err := r.getGracefulShutdownAndUpdatePodCheck(ctx, apiClient, svc, k8sSvcNamespace)
+			if err != nil {
+				r.Log.Error(err, "failed to get pod shutdown duration", "svc", svc.ServiceName)
+				errs = multierror.Append(errs, err)
 			}
-		} else {
+
+			// set requeue response, then continue to the next service instance
+			if podShutdownDuration > requeueAfter {
+				requeueAfter = podShutdownDuration
+			}
+			if podShutdownDuration > 0 {
+				continue
+			}
+
+			// If the service address is not in the Endpoints addresses, deregister it.
 			r.Log.Info("deregistering service from consul", "svc", svc.ServiceID)
-			_, err := apiClient.Catalog().Deregister(&api.CatalogDeregistration{
+			_, err = apiClient.Catalog().Deregister(&api.CatalogDeregistration{
```

Comment:
I think this is ok and not a concerning change in behavior, but wanted to spell out my understanding:

- This is within the scope of a single service (endpoints) being reconciled; that means the only time we expect heterogeneity between pods' graceful shutdown timings, is when deployments are being updated or in weirder pets-not-cattle pod treatment
- Even if one pod manages to artificially delay the deregistration of another's service instance, any pod that has finished terminating will be deregistered immediately
- Any pod without graceful shutdown configured is unaffected by this change

Reply:
I think this is a good summary with maybe one clarification: 
> Even if one pod manages to artificially delay the deregistration of another's service instance, any pod that has finished terminating will be deregistered immediately

I don't think in this case it will be deregistered immediately, but:
1. When the pod finishes normally, consul-server-connection-manager should clean up the ACL token immediately
1. Even if the pod is force-killed by k8s, the consul service instance will be `critical` health, so it should not be routable.
1. Any other endpoint reconciliation for that service would cause that instance to be deregistered prior to the `requeueAfter` time. 

Let me know if you see any problems with this approach.

Reply:
> Any other endpoint reconciliation for that service would cause that instance to be deregistered prior to the requeueAfter time.

Good clarification. I guess when I said "immediately", I meant "during the next reconcile", which I expect would follow that state transition (though maybe that's actually a behavioral difference between Endpoints and EndpointSlices). Given the other safeguards mentioned, this seems ok to me.

One other question I have about the `critical` health update: is there a possibility that status would flap on a subsequent health check of the service, if it isn't yet self-reporting as unhealthy due to shutdown?

Reply:
> One other question I have about the critical health update: is there a possibility that status would flap on a subsequent health check of the service, if it isn't yet self-reporting as unhealthy due to shutdown?

My thought here is that consul-k8s is the only source of health check information and this is only otherwise set during service registration. Since the pod should be removed from the endpoints object indefinitely, I would expect that we would never get back to a registration event that would toggle it. At least that's what I observed doing testing.